desc/objectives/_stability.py

```diff
@@ -367,3 +369,278 @@ def compute(self, params, constants=None):
+            jnp.pi * self._nturns,
+            self._nturns * self._nzetaperturn,
+        )
+
+        # set alpha/zeta0 grids
+        self._alpha = setdefault(
+            self._alpha, jnp.linspace(0, jnp.pi, 8) if eq.N != 0 else jnp.array(0.0)
+        )
+        self._zeta0 = setdefault(
+            self._zeta0, jnp.linspace(-0.5 * jnp.pi, 0.5 * jnp.pi, 15)
+        )
+        self._dim_f = 1
+        self._data_keys = ["ideal ball lambda"]
+
+        self._args = get_params(
+            self._iota_keys + self._len_keys + self._data_keys,
+            obj="desc.equilibrium.equilibrium.Equilibrium",
+            has_axis=False,
+        )
+
+        self._constants = {
+            "iota_transforms": iota_transforms,
+            "iota_profiles": iota_profiles,
+            "len_transforms": len_transforms,
+            "len_profiles": len_profiles,
+            "rho": self._rho,
+            "alpha": self._alpha,
+            "zeta": zeta,
+            "zeta0": self._zeta0,
+            "lam0": self._lam0,
+            "w0": self._w0,
+            "w1": self._w1,
+            "quad_weights": 1.0,
+        }
+        super().build(use_jit=use_jit, verbose=verbose)
+
+    def compute(self, params, constants=None):
+        """
+        Compute the ballooning stability growth rate.
+
+        Parameters
+        ----------
+        params : dict
+            Dictionary of equilibrium degrees of freedom, eg Equilibrium.params_dict
+        constants : dict
+            Dictionary of constant data, eg transforms, profiles etc. Defaults to
+            self.constants
+
+        Returns
+        -------
+        lam : ndarray
+            ideal ballooning growth rate.
+
+        """
+        eq = self.things[0]
+
+        if constants is None:
+            constants = self.constants
+        # we first compute iota on a uniform grid to get correct averaging etc.
+        iota_data = compute_fun(
+            eq,
+            self._iota_keys,
+            params=params,
+            transforms=constants["iota_transforms"],
+            profiles=constants["iota_profiles"],
+        )
+
+        len_data = compute_fun(
+            eq,
+            self._len_keys,
+            params=params,
+            transforms=constants["len_transforms"],
+            profiles=constants["len_profiles"],
+        )
+
+        # Now we compute theta_DESC for given theta_PEST
+        rho, alpha, zeta = constants["rho"], constants["alpha"], constants["zeta"]
+
+        # we prime the data dict with the correct iota values so we don't recompute them
+        # using the wrong grid
+        data = {
+            "iota": iota_data["iota"][0],
+            "iota_r": iota_data["iota_r"][0],
+            "shear": iota_data["shear"][0],
+            "a": len_data["a"],
+        }
+
+        grid = eq.get_rtz_grid(
+            rho,
+            alpha,
+            zeta,
+            coordinates="raz",
+            period=(np.inf, 2 * np.pi, np.inf),
```

Comment:
For the reasons discussed in #1180 , I think you want `period=(np.inf, np.inf, np.inf)`.

When you map coordinates with finite period $2\pi$ you are finding the values `mod_theta` = $\theta(\alpha, \zeta) mod (2 \pi)$ where $\theta(\alpha, \zeta)$ is the coordinate such that the physical location you are at $\alpha, \zeta$ can be computed by the relation $\alpha = \theta(\alpha, \zeta) + \lambda(\rho, \theta, \zeta) - \iota \zeta$.

If your intention is to only feed these $\theta$ values into functions that are periodic with $2\pi$, then this loss of information from modding is fine, since it's never used anyway. However, the compute functions for the stability quantities use `alpha_r` (through `g^aa` <- `grad(alpha)` <- `alpha_r`), which is not a periodic function of $\theta$. Hence, if we plotted the `alpha_r` used in this objective's computations, it would have jump discontinuities along moving along $\zeta$ holding the field line fixed. These jump discontinuities are not necessarily at points where $\alpha$ is close to $2\pi$. Hence, even functions that are periodic with $\alpha$ will be affected by these jump discontinuities.

Also, modding $\theta$ forces `grad(alpha)` to be bounded. In theory,  $\nabla \alpha$ should become unbounded as $\zeta \to \infty$ and $(\partial\alpha/\partial\rho)_{\theta, \zeta}$ should be continuous along the field line.

Reply:
(It's possible this had no effect if you aren't following the field line for enough toroidal transits for the modding to have an effect.)

Reply:
Sorry `alpha_r` is not affected. It's just `alpha`. And $\zeta$ is not modded so $\nabla \alpha$ does what is expected.